File: platform-common-api/src/main/java/org/sweetchips/platform/common/ContextLogger.java
package org.sweetchips.platform.common;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Locale;

public class ContextLogger {

    public static final int DEBUG = 3;
    public static final int INFO = 4;
    public static final int WARN = 5;
    public static final int ERROR = 6;

    private static final ContextLogger sDefault = new ContextLogger();

    public static ContextLogger getDefault() {
        return sDefault;
    }

    public void d(String tag, String msg) {
        System.out.println(String.format(Locale.getDefault(), "I/%s: %s", tag, msg));
    }

    public void d(String tag, Throwable e) {
        System.out.println(String.format(Locale.getDefault(), "I/%s: %s", tag, printStackTrace(e)));
    }

    public void i(String tag, String msg) {
        System.out.println(String.format(Locale.getDefault(), "D/%s: %s", tag, msg));
    }

    public void i(String tag, Throwable e) {
        System.out.println(String.format(Locale.getDefault(), "D/%s: %s", tag, printStackTrace(e)));
    }

    public void w(String tag, String msg) {
        System.err.println(String.format(Locale.getDefault(), "W/%s: %s", tag, msg));
    }

    public void w(String tag, Throwable e) {
        System.err.println(String.format(Locale.getDefault(), "W/%s: %s", tag, printStackTrace(e)));
    }

    public void e(String tag, String msg) {
        System.err.println(String.format(Locale.getDefault(), "E/%s: %s", tag, msg));
    }

    public void e(String tag, Throwable e) {
        System.err.println(String.format(Locale.getDefault(), "E/%s: %s", tag, printStackTrace(e)));
    }

    private static String printStackTrace(Throwable e) {
        StringWriter stringWriter = new StringWriter();
        e.printStackTrace(new PrintWriter(stringWriter));
        return stringWriter.toString();
    }
}

File: platform-common/src/main/kotlin/org/sweetchips/platform/common/ContextLogger.kt
package org.sweetchips.platform.common

import java.io.PrintWriter
import java.io.StringWriter

open class ContextLogger {

    enum class Level {

        DEBUG {
            override fun log(logger: ContextLogger, tag: String, msg: String) {
                logger.d(tag, msg)
            }
        },
        INFO {
            override fun log(logger: ContextLogger, tag: String, msg: String) {
                logger.i(tag, msg)
            }
        },
        WARN {
            override fun log(logger: ContextLogger, tag: String, msg: String) {
                logger.w(tag, msg)
            }
        },
        ERROR {
            override fun log(logger: ContextLogger, tag: String, msg: String) {
                logger.e(tag, msg)
            }
        };

        abstract fun log(logger: ContextLogger, tag: String, msg: String)
    }

    companion object {

        private fun Throwable.logStackTrace(): String {
            return StringWriter().also {
                printStackTrace(PrintWriter(it))
            }.toString()
        }
    }

    open fun d(tag: String, msg: String) {
        print("${Level.DEBUG}/$tag: $msg")
    }

    open fun d(tag: String, e: Throwable) {
        d(tag, e.logStackTrace())
    }

    open fun i(tag: String, msg: String) {
        print("${Level.INFO}/$tag: $msg")
    }

    open fun i(tag: String, e: Throwable) {
        i(tag, e.logStackTrace())
    }

    open fun w(tag: String, msg: String) {
        print("${Level.WARN}/$tag: $msg")
    }

    open fun w(tag: String, e: Throwable) {
        w(tag, e.logStackTrace())
    }

    open fun e(tag: String, msg: String) {
        print("${Level.ERROR}/$tag: $msg")
    }

    open fun e(tag: String, e: Throwable) {
        e(tag, e.logStackTrace())
    }

    open fun log(level: Level, tag: String, msg: String) {
        level.log(this, tag, msg)
    }

    open fun log(level: Level, tag: String, e: Throwable) {
        log(level, tag, e.logStackTrace())
    }
}